README.md
# bibliotek-o: A Framework for Modeling Bibliographic Metadata 

## What is bibliotek-o?

bibliotek-o is an ontology framework for modeling bibliographic metadata based on the [Library of Congress BIBFRAME ontology](http://id.loc.gov/ontologies/bibframe). It includes:

* The [bibliotek-o ontology](http://bibliotek-o.org/ontology/), which defines extensions and modifications to BIBFRAME and is intended to be used as a supplement to the core BIBFRAME ontology.
* A set of external ontology fragments (including BIBFRAME) intended for use with the bibliotek-o ontology.
* A data model and application profile - i.e., a set of recommended models and patterns for expressing bibliographic metadata using BIBFRAME, the bibliotek-o ontology, and other external ontologies.

bibliotek-o is a joint product of the Mellon Foundation-funded Linked Data for Libraries Labs (LD4L Labs) and Linked Data for Production (LD4P) projects. See the [LD4L website](http://ld4l.org), the [LD4L Labs public wiki](https://wiki.duraspace.org/pages/viewpage.action?pageId=77447730), and the [LD4P public wiki](https://wiki.duraspace.org/pages/viewpage.action?pageId=74515029) for more information about these projects.

## News

* **2017-07-05** Q2 release of Version 1.1.0 
* Watch for an announcement of the Q3 release and timeline.


## Repository contents

This repository contains the formal bibliotek-o specification, ontology and dataset recommendations for use with bibliotek-o (including the recommended BIBFRAME terms), and a set of human-readable documentation. 

* [`target-ontologies/bibliotek-o.owl`](target-ontologies/bibliotek-o.owl): OWL specification of bibliotek-o 
* [`target-ontologies/`](target-ontologies/): RDFS and OWL specifications of recommended ontology fragments, including BIBFRAME, for use with bibliotek-o
* [`reference-ontologies/`](reference-ontologies): Full specifications of these ontologies, for reference purposes
* [`xsd/bibliotek-o.datatypes.xsd`](xsd/bibliotek-o.datatypes.xsd): Datatypes defined by bibliotek-o
* [`datasets/`]: Datasets included in the bibliotek-o recommendation 
* [`doc/lode/`](doc/lode/): Human-readable documentation (generated by the [online LODE tool](http://www.essepuntato.it/lode)): 
* [`doc/overview/`](doc/overview/): Overview diagrams of the bibliotek-o application profile, in both vue and png formats
* [`doc/principles/`](doc/principles/): The foundational principles applied to the development of bibliotek-o
* [`doc/admin/`](doc/admin/): Documentation of administrative processes and procedures, such as change and release processes
* [`CHANGELOG.md`](CHANGELOG.md): Release change log
  
## Versioning

The following protocols are used to version and record changes to the ontology.
  
### Ontology versioning
* The use of OWL versioning predicates follows the [OWL 2 specification](https://www.w3.org/TR/owl2-syntax/#Ontology_IRI_and_Version_IRI).
* `owl:ontologyIRI` identifies the ontology; e.g., `http://bibliotek-o.org/ontology/`.
* `owl:versionIRI` identifies a particular version of the ontology; e.g., `http://bibliotek-o.org/1.1/ontology/`. 
* The `owl:ontologyIRI` redirects to the `owl:versionIRI` of the current version of the ontology.
* Previous versions of the ontology continue to be available at their `owl:versionIRI`.
* By importing or referencing terms from a particular `owl:versionIRI`, users are insulated from non-backward-compatible changes in newer published versions until they decide to upgrade.
* Version numbering convention (`MAJOR.MINOR.PATCH`):
    * `MAJOR`: non-backward-compatible modifications
    * `MINOR`: backward-compatible semantic modifications
    * `PATCH`: non-semantic modifications, such as changing an rdfs:label.
* The `owl:versionIRI` is updated for `MAJOR` and `MINOR` versions, but not `PATCH` versions. It thus includes only the `MAJOR` and `MINOR` version numbers, as in `http://bibliotek-o.org/1.1/ontology/`.
* To avoid an unnecessary proliferation of `MAJOR` versions, `owl:deprecated` is applied to terms slated for removal. A term deprecated in a `MINOR` version will be removed in the next `MAJOR` version. 
* `owl:versionInfo` provides a label containing the version number, formatted as "Version n.n.n"; e.g., "Version 1.1.0". This version number is also used to tag the repository, in this case `v1.1.0`. 

### Issuance and modification datetimes
* `dcterms:issued` is used on each ontology term, and on the ontology as a whole, to indicate datetime of first issuance.
* `dcterms:modified` is used on each ontology term, and on the ontology as a whole, to indicate last modification datetime. 
* Terms not modified since first issuance have the same `dcterms:issued` and `dcterms:modified` values.
* Datetime values are expressed in ISO-8601 format; e.g., "2017-04-22T01:30:00-04:00".
* The `dcterms` values do not include extraneous text, so that they are machine-readable without parsing. Change descriptions are provided by `skos:changeNote` (see following section).

### Change descriptions
* `skos:changeNote` is used to provide human-readable descriptions of term modifications. 
* One `skos:changeNote` is used per version. That is, if more than one change is made to the same term in the same version, all are recorded in the same change note. If multiple changes are made to the same term in different versions, they are recorded in multiple change notes.
* The change note also records the ontology version, so that modifications to a term can be traced through the affected versions of the ontology. The format used is, e.g., "Fix rdfs:label (v1.0.1)".
* A `skos:changeNote` could be applied to the ontology itself to record major, broad, or high-level changes affecting multiple terms.
* Changes are also recorded in the [project change log](CHANGELOG.md).

## Change and release process

Please refer to the [Change and Release Management](doc/admin/crm.md) documentation for details on this process.

## Additional bibliotek-o documentation 

* A [diagrammatic overview of the bibliotek-o application profile with an index of terms](http://bibliotek-o.org/overview.html) 
* An overview of the principles and best practices guiding the development of bibliotek-o (forthcoming)
* Discussion and diagrams of the principal models and patterns adopted by bibliotek-o (forthcoming)
* Mapping tables from current catalog formats to the ontology recommendation (forthcoming)
* Application profiles for original data modeling using the bibliotek-o recommendation (forthcoming)






RELEASES.md
#  bibliotek-o Change and Release Processes
 

This document articulates a structured process governing ongoing ontology changes and releases so that contributors, users, and other community members can plan and prepare for new versions.

## Change and release management team

* The change and release management (CRM) team is a dedicated team that oversees the change and release processes described below. 
* One team member assumes the role of team lead, which is primarily an administrative rather than decision-making role. The lead ensures that the change and release cycle proceeds according to schedule, calls meetings as necessary, and does initial triage of incoming issues.
* Team members and team lead may rotate periodically.
* Only CRM team members have write access to the repository (aside from GitHub organization owners). 
* Members of the CRM team must follow all procedures outlined below, rather than pushing directly to the repository, for proposing and making changes to the ontology.

## Issues 

GitHub issues are the vehicles for proposing change, requesting information or clarification, starting discussions, etc. Any interested community members are invited to submit issues as well as to watch individual issues or the entire repository. 

* Requests for modifications to bibliotek-o.owl, external ontologies (i.e., addition or removal of terms from the target files), datasets, datatypes, and artifacts such as documentation, are made through GitHub issues. These issues should be labelled `change request`.
* Contributors are encouraged to submit a pull request (see below) with an issue, while non-contributing users will submit issues as proposals for modification that may be assigned to contributors.
* Requests for clarification, modeling and implementation questions, discussion topics, etc. that are not direct proposals for change may also be submitted as GitHub issues. The issues should be labeled either `request for clarification` or `discussion`.  Such issues _may_ be reviewed and addressed outside the release cycle, but there is no commitment to do so.


### Standards for issue content
* Issues should be atomic. Multiple _interdependent_ modifications that are part of a single change should be submitted as a single issue, while multiple _unrelated_ changes should be submitted as separate issues. For example:
  * Atomic issue: "Define class Attribution, predicate hasAttribution, and inverse predicate isAttributionOf."
  * Non-atomic issue: "Define class Attribution and make predicate hasEquivalent symmetric."  
  * Where appropriate, the CRM team will break apart non-atomic issues into multiple issues and combine separate related issues into a single issue.
* Issues should include a rationale and use cases.
* Issues should include the ontology version number against which they are submitted.

## Pull requests

A pull request (PR) implements the changes proposed in an issue. Any interested community members are invited to submit PRs to the repository.

* PRs are made against the develop (default) branch.
* Each PR should address only one issue.
* A PR should include any modifications to the BIBFRAME and other target ontology files (i.e., adding or deleting terms from the fragments included in the repository) that are associated with the proposed change to bibliotek-o.
* PRs should not include regeneration of documentation, since this will be regenerated once prior to the release.
* The commit message should use a GitHub keyword to automatically close the related issue, such as `Fixes #5`. See [GitHub documentation on closing issues via commit messages](https://help.github.com/articles/closing-issues-via-commit-messages/).


### Standards for modifications to bibliotek-o.owl

* New term:
  * Include `dcterms:issued` and `dcterms:modified` assertions, with the same datetime value. Datetime values are expressed in ISO-8601 format; e.g., "2017-04-22T01:30:00-04:00".
* Modification of an existing term:
  * Update the `dcterms:modified` value.
  * Add a `skos:changeNote` describing the change, leaving a set of parentheses where the new version number will be filled in at release time; For example, "Fix rdfs:label ()".   
* Term deletion
  * Terms should be marked `owl:deprecated` rather than deleted. The CRM team will determine, based on the semantics of the change and whether the next release is a MAJOR, MINOR, or PATCH release, whether the term can be deleted rather than deprecated.
  * Update the `dcterms:modified` value and add a `skos:changeNote` as for modified terms.
* All annotation properties should have an `xml:lang` value. The value used for English is `"en"` (i.e., the variant of English is not specified). 

## Change cycle

The change cycle operates on a quarterly basis. Each new change cycle begins with the announcement of deadlines associated with the release, and ends with the release of a new version (assuming there are changes to release). Steps in the change cycle are outlined below.

### Initial community announcement 
An announcement of the date of next release and associated deadlines is posted at the top of this README.

### Issue review
* Immediately following the issue submission deadline (see timeline below), the CRM team begins to review open issues.
* The team lead performs initial triage and assigns issues for review to team members as appropriate. Team members may reassign an issue to another team member.
* Each issue receives a recommendation from the assigned team member to accept for the current cycle, reject, or defer, and presents the recommendation to the team. 
* Recommendations are reviewed by the entire team and resolves any differences. Meetings are called as desired.
* Issues are either accepted for the current cycle, rejected, or deferred. 

site/js/myjs.js
angular
.module('myApp',['ngSanitize'])
.controller('myCtrl', ['$scope', function($scope) {
        $scope.homePage = function() {
          $scope.homepageView = true;
          $scope.designView = false;
          $scope.patternView = false;
        };
        $scope.readPatterns = function() {
          $scope.homepageView = false;
          $scope.designView = false;
          $scope.patternView = true;

          d3.json("data/models.json", function(error, data) {
            $scope.processModlesJSONData(data);
          });

        };
        $scope.readJSON = function() {
          $scope.homepageView = false;
          $scope.patternView = false;
          $scope.designView = true;
          d3.json("data/modules.json", function(error, data) {
            $scope.processJSONData(data);
          });
        };

        $scope.processJSONData = function(data) {
          console.log("process json data");
          $scope.ontologies = data;
        };
        $scope.processModlesJSONData = function(data) {
          console.log("process models json data");
          $scope.models = data;
        };

        $scope.readJSON();
        $scope.readPatterns();
        $scope.homePage();

        $("#pop").on("click", function() {
            $('#imagepreview').attr('src', $('#imageresource').attr('src')); // here asign the image to the modal when the user click the enlarge link
            $('#imagemodal').modal('show'); // imagemodal is the id attribute assigned to the bootstrap modal, then i use the show function
        });

}]);


tools/fix-anchors.py
#!/usr/bin/env python
"""Fix up anchors in the BibX ontology document so that terms resolve.

The output of LODE is an HTML document with meaningless anchors which
means that we can't readily make the ontology URIs nicely resolve the
the appropriate definitions. This script goes through the HTML looking
for terms in the ontology and rewriting anchors to be the local part of
the URI.

Note that if there are cases where one ontology term (e.g. 
http://bibx.org/ontology/note ) is described in multiple places
within the ontology, the rewrite of anchors takes the first occurrence
in the HTML as the anchor to change.

Simeon Warner - 2016-01-22
"""

import re

html = open('../doc/LODE/bibx.html','r').read()
prefix = 'http://bibx.org/ontology/'

## Pass 1 - find anchors to change
terms = {}
seen_anchors = set()
for m in re.findall(r'''<a href="#([0-9a-f]+)" title="([^"]+)"''',html):
    anchor = m[0]
    uri = m[1]
    # Only want URIs in this namespace
    lmatch = re.match(prefix+r'''(\w+)$''', uri)
    if (lmatch):
        term = lmatch.group(1)
        if (term in terms):            
            if (terms[term] != anchor and anchor not in seen_anchors):
                ns = prefix if (not lmatch.group(1)) else prefix + lmatch.group(1)
                print "Ignoring additional anchor %s, already have %s for %s%s" % (anchor, terms[term], ns, term)
                seen_anchors.add(anchor)
        else:
            terms[term] = anchor

## Pass 2 - implement changes
for term, anchor in terms.items():
    print "Replacing anchor %s with %s" % (anchor, term)
    html = re.sub(' id="'+anchor+'"',' id="'+term+'"',html)
    html = re.sub(' href="#'+anchor+'"',' href="#'+term+'"',html)

open('../doc/LODE/ontology.html','w').write(html)


tools/fix-lode-anchors.py
#!/usr/bin/env python
"""Fix up anchors in the bibliotek-o LODE ontology document so that terms resolve.

The output of LODE is an HTML document with meaningless anchors which
means that we can't readily make the ontology URIs nicely resolve the
the appropriate definitions. This script goes through the HTML looking
for terms in the ontology and rewriting anchors to be the local part of
the URI.

Note that if there are cases where one ontology term is described in multiple 
places within the ontology, the rewrite of anchors takes the first occurrence
in the HTML as the anchor to change.

Before running the script, change the prefix value to the versionIRI of the
ontology: e.g., http://bibliotek-o.org/1.2/ontology for version 1.2.x.

After running this script, change the href on the "Ontology Source" link
to http://biblioteko-org/ontology.owl.

Simeon Warner - 2016-01-22
"""

import re

html = open('../doc/lode/bibliotek-o.html','r').read()
prefix = 'http://bibliotek-o.org/1.1/ontology/'

## Pass 1 - find anchors to change
terms = {}
seen_anchors = set()
for m in re.findall(r'''<a href="#([0-9a-f]+)" title="([^"]+)"''',html):
    anchor = m[0]
    uri = m[1]
    # Only want URIs in this namespace
    lmatch = re.match(prefix+r'''(\w+)$''', uri)
    if (lmatch):
        term = lmatch.group(1)
        if (term in terms):            
            if (terms[term] != anchor and anchor not in seen_anchors):
                ns = prefix if (not lmatch.group(1)) else prefix + lmatch.group(1)
                print "Ignoring additional anchor %s, already have %s for %s%s" % (anchor, terms[term], ns, term)
                seen_anchors.add(anchor)
        else:
            terms[term] = anchor

## Pass 2 - implement changes
for term, anchor in terms.items():
    print "Replacing anchor %s with %s" % (anchor, term)
    html = re.sub(' id="'+anchor+'"',' id="'+term+'"',html)
    html = re.sub(' href="#'+anchor+'"',' href="#'+term+'"',html)

open('../doc/lode/ontology.html','w').write(html)
